### PR TITLE
Add nix_dump_cst_json

### PR DESCRIPTION
**Pull request by adrianparvino**
Supersedes #20 and #41

### CODE REVIEW

File: nixui/options/ast.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import collections
 import parsimonious
```
Comment (lapp0) [inline nixui/options/ast.py:2]: unused import

File: nixui/options/ast.py [hunk 2]
```diff
@@ -34,80 +35,30 @@ def get_full_node_string(node):
...
-
 def parse_ast_str(ast_str):
-    # parsing the dumped ast string
-    tree = AST_DUMP_GRAMMAR.parse(ast_str)
-    root_node = NixVisitor().visit(tree)
-    assert len(root_node) == 1
-    return root_node[0]
+    return parse_ast_dict_node_or_token(json.loads(ast_str))
 
+def parse_ast_dict_node_or_token(d):
```
Comment (lapp0) [inline nixui/options/ast.py:53]: Nit: this should be defined after `Node`, `Token` and `NumRange` are

Nit: We should call it `cst` not `ast` if the rust program returns a concrete syntax tree.

File: nix_dump_cst_json/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "nix_dump_cst_json"
+version = "0.1.0"
+edition = "2018"
```
Comment (lapp0) [inline nix_dump_cst_json/Cargo.toml:4]: Why 2018?

**Review by lapp0 — CHANGES_REQUESTED** [review]
looks good, a few minor changes and clarifications

File: nixui/options/ast.py (hunk 2 above)
Comment (lapp0) [reply nixui/options/ast.py:53]: or instead of cst, everywhere we could just say `syntax_tree`.

File: nix_dump_cst_json/Cargo.toml (hunk 3 above)
Comment (adrianparvino) [reply nix_dump_cst_json/Cargo.toml:4]: `edition` here is the Rust edition, and the latest seems to be 2018.

File: nixui/options/ast.py (hunk 2 above)
Comment (adrianparvino) [reply nixui/options/ast.py]: https://github.com/nix-gui/nix-gui/pull/42/commits/01384311f0f41db6260f552da567370fff2f7acc

**Comment by adrianparvino** [comment]
I think there's still something broken with UTF-8 decoding.

File: nixui/options/ast.py (hunk 1 above)
Comment (adrianparvino) [reply nixui/options/ast.py]: https://github.com/nix-gui/nix-gui/pull/42/commits/a5336d70cb1e219a7f579f6badf23c99fb90ff02

**Comment by lapp0** [comment]
> I think there's still something broken with UTF-8 decoding.
> 
> I _think_ that it will break if we have something along the lines of
> `services.foo."⏪bar" = ...` because the (line, column) -> byte offset mapping is not robust against this.
> 
> Currently, the line mapping to byte offset works,
> https://github.com/nix-gui/nix-gui/blob/01384311f0f41db6260f552da567370fff2f7acc/nixui/options/parser.py#L32
> 
> 
> however column does not.
> https://github.com/nix-gui/nix-gui/blob/01384311f0f41db6260f552da567370fff2f7acc/nixui/options/parser.py#L34
> 
> 
> Current tests won't detect this because all instances of these are located in comments. In the earlier example, there'll be an offset error when trying to access `...` because `⏪` only advances by one byte offset (as it's assumed that one column offset is one byte offset.)

I wonder if there's a way to apply a transformation to fix this? Perhaps when calculating position run `str.decode('unicode_escape')`?

**Pull request merged**